metabase/metabase
# feat(sdk): add Next.js compatibility to embedding cli

Closes https://github.com/metabase/metabase/issues/50096

### Description

Adds in basic Next.js compatibility to the Embedding CLI, as follows:

- Uses the experimental `/nextjs` import for Next.js projects when a Next.js project is detected.
- Adds a `use client` to the generated components when necessary
- Creates an `/analytics-demo` route in their router of choice (App or Pages)
- Creates an example root layout or custom app if they don't already have one, or prints out how to setup the demo CSS stylesheets and sample providers.

### Out of scope / Known Issues

- "Dashboard 11 not found" when switching between dashboards. This is fixed in https://github.com/metabase/metabase/pull/53452
- The Next.js app briefly shows "the component requires the `<MetabaseProvider>`" for [a split second](https://www.loom.com/share/23d0e57a1ef8443b8a3577c81d27afb8?sid=f1a6435a-07e6-420c-aa04-ae8c7bb8fc9d) before disappearing when **switching between themes**. This looks like be a limitation of SSR or the experimental compatibility layer.
- The "Loading..." message being unthemed - has to be themed on the SDK side, as I think it is part of the compatibility layer's loader.
- Adding TypeScript type annotations is out of scope. Currently the generated snippets does not have TypeScript type annotations, and will need a bit of work by the user to add the annotations.
- Adding an API route for `/metabase/sso` and `/user-switch` in case they choose to do multi-tenancy is out of scope. Currently this is still a separate Express.js server. I added a message that they have to convert the API routes to API routes and a link to the sample app so they can set it up.

### Testing Dimensions

- Has TypeScript in project (i.e. has `tsconfig.json` or the `typescript` library): yes / no
- Is a Next.js project (i.e. has `next.config.ts` or `next.config.js` or has the `next` library): yes / no
- Has a valid Pro/Enterprise license been applied: yes / no
- Next.js Router type (i.e. does it have `pages` or `app` directory): Pages Router vs App Router
- Does the project has an existing Root Layout (App router) or Oustom App (Pages router): yes / no

### How to verify

- [Create a new Next.js project](https://nextjs.org/learn-pages-router/basics/create-nextjs-app/setup)
- Run the CLI in it. Use the following command: `docker rm -f metabase-enterprise-embedding && yarn add file:../metabase/resources/embedding-sdk && npx @metabase/embedding-sdk-react start`
  - If you are using Yarn 2, you need to specify the prefix to link it: `docker rm -f metabase-enterprise-embedding && yarn add "@metabase/embedding-sdk-react@file:../metabase/resources/embedding-sdk" && npx @metabase/embedding-sdk-react start` 
- It should use the experimental Next.js layer with your app.
- It should generate an `/analytics-demo` endpoint, and they can visit `/analytics-demo` on their Next.js app.
- The components should work with Next.js.

### Demo

![CleanShot 2568-02-10 at 20 25 45@2x](https://github.com/user-attachments/assets/18a83c59-234a-42cf-81c2-e5bfcdf23ffd)
